File: ScalaWeb/src/Data/BaseData.scala
package Data

import MySql.{DataBase, DataParameterList, TDataBase, TDataParameterList}
import Utility.{AppSettings, Common}

/**
  * Created by BianzhaiWang on 2017/1/19.
  */
class BaseData {
  var DBAccess: TDataBase = null

  var TableName: String = null

  var PrimaryKey: String = null

  Init()

  private def Init() {
    if (DataBase.ConnectionString == null) {
      DataBase.ConnectionString = AppSettings.ConnectionString
    }
    this.DBAccess = new DataBase()
  }

  def SelectEntities[T](cls: Class[_], sql: String, parameterList: TDataParameterList): List[T] = {
    return this.DBAccess.ExceSelectTo(cls, sql, parameterList)
  }

  def SelectEntities[T](cls: Class[_], selectFieldList: List[String], whereNameValues: Map[String, Any]): List[T] = {
    val parameterList: TDataParameterList = new DataParameterList()

    val fieldList = whereNameValues.keys.map(m => String.format("%s=@%s ", m, m)).toList
    parameterList.Set(whereNameValues)

    val selectFieldSql = if (selectFieldList == null) "*" else Common.StringJoin(",", selectFieldList)
    val sql = String.format("select %s from %s where %s", selectFieldSql, this.TableName, Common.StringJoin(" and ", fieldList))

    return this.SelectEntities(cls, sql, parameterList)
  }

  def SelectEntities[T](cls: Class[_]): List[T] = {
    val sql = String.format("select * from %s", this.TableName)
    return this.SelectEntities(cls, sql, null)
  }

  def SelectEntity[T](cls: Class[_], sql: String, parameterList: TDataParameterList): T = {
    return Common.GetFirstOrDefault(cls, this.SelectEntities(cls, sql, parameterList))
  }

  def SelectEntity[T](cls: Class[_], selectFieldList: List[String], whereNameValues: Map[String, Any]): T = {
    return Common.GetFirstOrDefault(cls, this.SelectEntities(cls, selectFieldList, whereNameValues))
  }

  def Insert(nameValues: Map[String, Any]): Boolean = {
    val nameList = nameValues.keys.toList

    val sql = String.format("insert %s (%s) values (%s)", this.TableName, Common.StringJoin(",", nameList),
      Common.StringJoin(",", nameList.map(s => String.format("@%s", s)).toList))

    val parameterList: TDataParameterList = new DataParameterList()
    parameterList.Set(nameValues)

    return this.ExceNoQuery(sql, parameterList) == 1
  }

  def ExceNoQuery(sql: String, parameterList: TDataParameterList): Int = {
    return this.DBAccess.ExceNoQuery(sql, parameterList)
  }

  def Update(nameValues: Map[String, Any], whereSql: String, parameterList: TDataParameterList): Boolean = {
    val parameterList2 = if (parameterList == null) new DataParameterList() else parameterList

    val fieldList = nameValues.keys.map(m => String.format("%s=@%s", m, m)).toList
    val sql = String.format("udpate %s set %s %s", this.TableName, Common.StringJoin(",", fieldList), whereSql)
    parameterList2.AddMap(nameValues)

    return this.ExceNoQuery(sql, parameterList2) == 1
  }

  def Update(nameValues: Map[String, Any], whereNameValues: Map[String, Any]): Boolean = {
    val parameterList: TDataParameterList = new DataParameterList()
    parameterList.Set(whereNameValues)

    val fieldList = whereNameValues.keys.map(m => String.format("%s=@%s", m, m)).toList
    val whereSql = String.format("where %s", Common.StringJoin(" and ", fieldList))

    return this.Update(nameValues, whereSql, parameterList)
  }

  def Update(nameValues: Map[String, Any], primaryKeyValue: Any): Boolean = {
    var whereNameValues = Map[String, Any]()
    whereNameValues += (this.PrimaryKey -> primaryKeyValue)
    return this.Update(nameValues, whereNameValues)
  }

  def Delete(whereNameValues: Map[String, Any]): Boolean = {
    val parameterList: TDataParameterList = new DataParameterList()
    parameterList.Set(whereNameValues)

    val fieldList = whereNameValues.keys.map(m => String.format("%s=@%s", m, m)).toList
    val sql = String.format("delete from %s where %s", this.TableName, Common.StringJoin(" and ", fieldList))

    return this.ExceNoQuery(sql, parameterList) > 0
  }

  def Delete(primaryKeyValue: Any): Boolean = {
    var whereNameValues = Map[String, Any]()
    whereNameValues += (this.PrimaryKey -> primaryKeyValue)
    return this.Delete(whereNameValues)
  }
}

File: WebFramework3/WebApi/src/Data/BaseData.java
package Data;

import LambdaInterface.IExceptionHandle;
import Oracle.DataBase;
import Oracle.DataParameterList;
import Oracle.IDataBase;
import Oracle.IDataParameterList;
import Utility.AppSettings;
import Utility.Common;

import java.util.*;
import java.util.stream.Collectors;

/**
 * Created by Bianshuai on 2017/1/11.
 */
public class BaseData {
    public IDataBase DBAccess = null;

    public String TableName = null;

    public String PrimaryKey = null;

    public IExceptionHandle ExceptionHandle = null;

    public BaseData() {
        this.Init();
    }

    private void Init() {
        if (DataBase.ConnectionString == null) {
            DataBase.ConnectionString = AppSettings.ConnectionString;
            DataBase.User = AppSettings.DbUser;
            DataBase.Password = AppSettings.DbPassword;
        }
        this.DBAccess = new DataBase();
    }

    public BaseData(IExceptionHandle exHandle) {
        this.Init();
        this.ExceptionHandle = exHandle;
    }

    public <T> List<T> SelectEntities(Class<T> cls, String sql, IDataParameterList parameterList) {
        try {
            return this.DBAccess.ExceSelectTo(cls, sql, parameterList);
        } catch (Exception ex) {
            this.ExHandling(ex);
            return null;
        }
    }

    private void ExHandling(Exception ex) {
        if (this.ExceptionHandle != null) {
            this.ExceptionHandle.Handling(ex);
        }
    }

    public <T> List<T> SelectEntities(Class<T> cls, List<String> selectFieldList, Map<String, Object> whereNameValues) {
        IDataParameterList parameterList = new DataParameterList();

        List<String> fieldList = whereNameValues.keySet().stream().map(m -> String.format("%s=@%s ", m, m)).collect(Collectors.toList());
        parameterList.Set(whereNameValues);

        String selectFieldSql = selectFieldList == null ? "*" : String.join(",", selectFieldList);

        String sql = String.format("select %s from %s where %s", selectFieldSql, this.TableName, String.join(" and ", fieldList));

        return this.SelectEntities(cls, sql, parameterList);
    }

    public <T> List<T> SelectEntities(Class<T> cls) {
        String sql = String.format("select * from %s", this.TableName);
        return this.SelectEntities(cls, sql, null);
    }

    public <T> T SelectEntity(Class<T> cls, String sql, IDataParameterList parameterList) {
        return Common.GetFirstOrDefault(cls, this.SelectEntities(cls, sql, parameterList));
    }

    public <T> T SelectEntity(Class<T> cls, List<String> selectFieldList, Map<String, Object> whereNameValues) {
        return Common.GetFirstOrDefault(cls, this.SelectEntities(cls, selectFieldList, whereNameValues));
    }

    public boolean Insert(Map<String, Object> nameValues) {
        Set<String> nameList = nameValues.keySet();

        String sql = String.format("insert %s (%s) values (%s)", this.TableName, String.join(",", nameList),
                String.join(",", nameList.stream().map(s -> String.format("@%s", s)).collect(Collectors.toList())));

        IDataParameterList parameterList = new DataParameterList();
        parameterList.Set(nameValues);

        return this.ExceNoQuery(sql, parameterList) == 1;
    }

    public int ExceNoQuery(String sql, IDataParameterList parameterList) {
        try {
            return this.DBAccess.ExceNoQuery(sql, parameterList);
        } catch (Exception ex) {
            this.ExHandling(ex);
            return -1;
        }
    }

    public boolean Update(Map<String, Object> nameValues, String whereSql, IDataParameterList parameterList) {
        if (parameterList == null) {
            parameterList = new DataParameterList();
        }

        List<String> fieldList = nameValues.keySet().stream().map(m -> String.format("%s=@%s", m, m)).collect(Collectors.toList());
        parameterList.AddMap(nameValues);

        String sql = String.format("update %s set %s %s", this.TableName, String.join(",", fieldList), whereSql);

        return this.ExceNoQuery(sql, parameterList) == 1;
    }

    public boolean Update(Map<String, Object> nameValues, Map<String, Object> whereNameValues) {
        IDataParameterList parameterList = new DataParameterList();
        parameterList.Set(whereNameValues);

        List<String> fieldList = whereNameValues.keySet().stream().map(m -> String.format("%s=@%s", m, m)).collect(Collectors.toList());

        String whereSql = String.format("where %s", String.join(" and ", fieldList));

        return this.Update(nameValues, whereSql, parameterList);
    }

    public boolean Update(Map<String, Object> nameValues, Object primaryKeyValue) {
        Map<String, Object> whereNameValues = new HashMap<>();
        whereNameValues.put(this.PrimaryKey, primaryKeyValue);
        return this.Update(nameValues, whereNameValues);
    }

    public boolean Delete(Map<String, Object> whereNameValues) {
        IDataParameterList parameterList = new DataParameterList();
        parameterList.Set(whereNameValues);

        List<String> fieldList = whereNameValues.keySet().stream().map(m -> String.format("%s=@%s", m, m)).collect(Collectors.toList());

        String sql = String.format("delete from %s where %s", this.TableName, String.join(" and ", fieldList));

        return this.ExceNoQuery(sql, parameterList) > 0;
    }

    public boolean Delete(Object primaryKeyValue) {
        Map<String, Object> whereNameValues = new HashMap<>();
        whereNameValues.put(this.PrimaryKey, primaryKeyValue);
        return this.Delete(whereNameValues);
    }
}
